src/utils/taint.js

```diff
@@ -0,0 +1,398 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import { CompilerDiagnostic, FatalError } from "../errors.js";
+import type { Realm } from "../realm.js";
+import type { Descriptor, PropertyBinding, ObjectKind } from "../types.js";
+import {
+  DeclarativeEnvironmentRecord,
+  FunctionEnvironmentRecord,
+  ObjectEnvironmentRecord,
+  GlobalEnvironmentRecord,
+} from "../environment.js";
+import {
+  BoundFunctionValue,
+  ProxyValue,
+  SymbolValue,
+  AbstractValue,
+  EmptyValue,
+  FunctionValue,
+  PrimitiveValue,
+  Value,
+  ObjectValue,
+  NativeFunctionValue,
+} from "../values/index.js";
+import { IsDataDescriptor, TestIntegrityLevel } from "../methods/index.js";
+import type { BabelNodeExpression } from "babel-types";
+import invariant from "../invariant.js";
+
+import { ResidualHeapInspector } from "../serializer/ResidualHeapInspector.js";
+
+function emitAllProperties(realm: Realm, O: ObjectValue) {
+  invariant(realm.generator, "a generator must exist while tracking a pure function");
+  let generator = realm.generator;
+  invariant(
+    O !== realm.$GlobalObject,
+    "we should not be able to get here because the global object is never mutated in a pure function."
+  );
+  // Temporary hack
+  let insp = new ResidualHeapInspector(realm, (null: any));
+  for (let [P, propertyBinding] of O.properties) {
+    if (insp.canIgnoreProperty(O, P)) {
+      continue;
+    }
+    let desc = propertyBinding.descriptor;
+    if (desc === undefined) continue; // deleted
+    if (IsDataDescriptor(realm, desc) && desc.configurable && desc.enumerable && desc.writable) {
+      let descValue = desc.value || realm.intrinsics.undefined;
+      invariant(descValue instanceof Value);
+      generator.emitPropertyAssignment(O, P, descValue);
+    } else {
+      generator.emitDefineProperty(O, P, desc);
```

Comment:
It this safe to do if the property is not configurable?

Reply:
This is emitted on a newly created object so it won't be. Basically it serializes its initial state here, but it can have multiple initial states if there are several branches. I'm not very happy with this strategy though.

(If anything, the property assignment above isn't safe if `Object.prototype` has properties that are non-writable, but we assume they don't.)